### PR TITLE
etcdserver: support sorted range requests in v3 api

### LINKED ISSUE #4166: support sort options for RangeRequest
The RangeRequest API supports limiting the number of keys returned. However, there's no way to control the order of the returned key/value results. For some algorithms this can necessitate reading in the entire range to process only a single key.

At minimum, a sorted RangeRequest would need to handle:
* Sort Attribute (key, mod rev, create rev, and maybe value)
* Order (ascending or descending)

Some use cases:
* Queues: sort by revision id ascending
* Priority queues: sort by key ascending
* Locks: sort by revision id, ascending and descending

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #4166

### CODE REVIEW

File: etcdserver/etcdserverpb/rpc.proto [hunk 1]
```diff
@@ -78,6 +78,21 @@ message RangeRequest {
   // if the revision has been compacted, ErrCompaction will be returned in
   // response.
   int64 revision = 4;
+
+  enum SortOrder {
+	NONE = 0; // default, no sorting
+	ASCEND = 1; // lowest key first
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: probably lowest value first (we can compare key,value of key, version, create, mod etc. not just key)

File: etcdserver/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -78,6 +78,21 @@ message RangeRequest {
...
+	DESCEND = 2; // highest key first
+  }
+  SortOrder sort_order = 5;
+  enum SortAttr {
+	KEY = 0;
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: we called this `CompareTarget`. We need to change to `sortTarget` or change `compareTarget` to `CompareAttr`? No strong opinion here.

File: etcdserver/etcdserverpb/rpc.proto [hunk 3]
```diff
@@ -78,6 +78,21 @@ message RangeRequest {
...
   // response.
   int64 revision = 4;
+
+  enum SortOrder {
+	NONE = 0; // default, no sorting
+	ASCEND = 1; // lowest key first
+	DESCEND = 2; // highest key first
+  }
+  SortOrder sort_order = 5;
+  enum SortAttr {
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: put the enum declarations together on the top?

File: etcdserver/v3demo_server.go [hunk 4]
```diff
@@ -181,6 +182,45 @@ func applyPut(txnID int64, kv dstorage.KV, p *pb.PutRequest) (*pb.PutResponse, e
...
+}
+func (s *kvSort) Len() int { return len(s.kvs) }
+
+type kvSortKey struct{ *kvSort }
+
+func (s *kvSortKey) Less(i, j int) bool {
+	return bytes.Compare(s.kvs[i].Key, s.kvs[j].Key) < 0
+}
+
+type kvSortVersion struct{ *kvSort }
```
Comment (xiang90) [inline etcdserver/v3demo_server.go]: kvSortByVersion

File: etcdserver/v3demo_server.go [hunk 5]
```diff
@@ -181,6 +182,45 @@ func applyPut(txnID int64, kv dstorage.KV, p *pb.PutRequest) (*pb.PutResponse, e
...
+
+func (s *kvSortVersion) Less(i, j int) bool {
+	return (s.kvs[i].Version - s.kvs[j].Version) < 0
+}
+
+type kvSortCreate struct{ *kvSort }
+
+func (s *kvSortCreate) Less(i, j int) bool {
```
Comment (xiang90) [inline etcdserver/v3demo_server.go]: kvSortByCreate

File: etcdserver/v3demo_server.go [hunk 6]
```diff
@@ -181,6 +182,45 @@ func applyPut(txnID int64, kv dstorage.KV, p *pb.PutRequest) (*pb.PutResponse, e
...
+	return (s.kvs[i].CreateRevision - s.kvs[j].CreateRevision) < 0
+}
+
+type kvSortMod struct{ *kvSort }
```
Comment (xiang90) [inline etcdserver/v3demo_server.go]: kvSortByMod

File: etcdserver/v3demo_server.go [hunk 7]
```diff
@@ -181,6 +182,45 @@ func applyPut(txnID int64, kv dstorage.KV, p *pb.PutRequest) (*pb.PutResponse, e
...
+
+func (s *kvSortMod) Less(i, j int) bool {
+	return (s.kvs[i].ModRevision - s.kvs[j].ModRevision) < 0
+}
+
+type kvSortValue struct{ *kvSort }
```
Comment (xiang90) [inline etcdserver/v3demo_server.go]: kvSortByValue

**Comment by xiang90** [comment]
LGTM. Have you tested it manually? (I assume we will cover this by integration test soon)

File: etcdserver/etcdserverpb/rpc.proto (hunk 2 above)
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: went with `SortTarget`

File: etcdserver/etcdserverpb/rpc.proto (hunk 1 above)
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: using "lowest target value"

**Comment by heyitsanthony** [comment]
Still untested. Will hold off on merge until then

**Comment by xiang90** [comment]
@heyitsanthony How about adding a few flags into etcdv3 range command. And then we can easily manually test it?

**Comment by heyitsanthony** [comment]
Sure, support for that should be there anyway. Will add.

**Comment by heyitsanthony** [comment]
Added support for limits, ordering, and sort targets to etcdctlv3. Tested on a few kv pairs, seems to work as intended. /cc @xiang90

File: etcdctlv3/command/range_command.go [hunk 8]
```diff
@@ -48,13 +58,40 @@ func rangeCommandFunc(cmd *cobra.Command, args []string) {
...
+	switch {
+	case sortOrder == "ASCEND":
+		sortByOrder = pb.RangeRequest_ASCEND
+	case sortOrder == "DESCEND":
+		sortByOrder = pb.RangeRequest_DESCEND
+	}
+
+	sortByTarget := pb.RangeRequest_KEY
+	switch {
+	case sortTarget == "CREATE":
```
Comment (xiang90) [inline etcdctlv3/command/range_command.go]: do we really want to differentiate upper and lower letters? Probably we can just do ToUpperLetters for the input flag?

File: etcdctlv3/command/range_command.go [hunk 9]
```diff
@@ -23,13 +23,23 @@ import (
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 )
 
+var (
+	rangeLimit int
+	sortOrder  string
```
Comment (xiang90) [inline etcdctlv3/command/range_command.go]: rangeSortOrder?

File: etcdctlv3/command/range_command.go [hunk 10]
```diff
@@ -23,13 +23,23 @@ import (
...
+	sortTarget string
```
Comment (xiang90) [inline etcdctlv3/command/range_command.go]: rangeSortTarget?

File: etcdctlv3/command/range_command.go [hunk 11]
```diff
@@ -48,13 +58,40 @@ func rangeCommandFunc(cmd *cobra.Command, args []string) {
 	if err != nil {
 		ExitWithError(ExitError, err)
 	}
+
+	sortByOrder := pb.RangeRequest_NONE
+	switch {
+	case sortOrder == "ASCEND":
+		sortByOrder = pb.RangeRequest_ASCEND
+	case sortOrder == "DESCEND":
+		sortByOrder = pb.RangeRequest_DESCEND
```
Comment (xiang90) [inline etcdctlv3/command/range_command.go]: we should probably return error on default. (bad flag value.)

File: etcdctlv3/command/range_command.go [hunk 12]
```diff
@@ -48,13 +58,40 @@ func rangeCommandFunc(cmd *cobra.Command, args []string) {
...
+	case sortTarget == "CREATE":
+		sortByTarget = pb.RangeRequest_CREATE
+	case sortTarget == "KEY":
+		sortByTarget = pb.RangeRequest_KEY
+	case sortTarget == "MODIFY":
+		sortByTarget = pb.RangeRequest_MOD
+	case sortTarget == "VALUE":
+		sortByTarget = pb.RangeRequest_VALUE
+	case sortTarget == "VERSION":
+		sortByTarget = pb.RangeRequest_VERSION
```
Comment (xiang90) [inline etcdctlv3/command/range_command.go]: exit with error on default case?

File: etcdctlv3/command/range_command.go [hunk 13]
```diff
@@ -48,13 +58,40 @@ func rangeCommandFunc(cmd *cobra.Command, args []string) {
...
 	}
 	kv := pb.NewKVClient(conn)
-	req := &pb.RangeRequest{Key: key, RangeEnd: rangeEnd}
-
+	req := &pb.RangeRequest{
+		Key:        key,
+		RangeEnd:   rangeEnd,
+		SortOrder:  sortByOrder,
+		SortTarget: sortByTarget,
+		Limit:      int64(rangeLimit)}
```
Comment (xiang90) [inline etcdctlv3/command/range_command.go]: new line for `}`?

File: etcdserver/etcdserverpb/rpc.proto [hunk 14]
```diff
@@ -78,6 +91,9 @@ message RangeRequest {
   // if the revision has been compacted, ErrCompaction will be returned in
   // response.
   int64 revision = 4;
+
+  SortOrder sort_order = 5;
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: add doc string for sort_order and sort_target?

**Comment by xiang90** [comment]
LGTM after addressing the issues mentioned above. Thanks!

File: etcdctlv3/command/range_command.go (hunk 11 above)
Comment (heyitsanthony) [inline etcdctlv3/command/range_command.go]: Reuse `ErrBadArgs` or should I create `ErrBadFlag`?

Comment (xiang90) [inline etcdctlv3/command/range_command.go]: probably ErrBadFlag. But I do not want to create so many error types either. Can you check what other command line tool does? Like curl.

Comment (heyitsanthony) [inline etcdctlv3/command/range_command.go]: curl gives:
1 for `curl foo://foo` (bad protocol)
2 for `curl --foo`, `curl` (bad args, bad flags)
4 for `curl --engine foo http://example.com` (feature not supported)
6 for `curl foo` (can't resolve)
So something like `ErrBadFeature` seems closest to what curl is doing

Comment (xiang90) [inline etcdctlv3/command/range_command.go]: @heyitsanthony (tried curl engine) Sounds good to me.

**Comment by heyitsanthony** [comment]
ok all fixed, will merge on greenlight

File: etcdctlv3/command/error.go [hunk 15]
```diff
@@ -27,6 +27,7 @@ const (
 	ExitError
 	ExitBadConnection
 	ExitInvalidInput // for txn, watch command
+	ExitBadFeature   // provided a valid flag with a bad value
```
Comment (xiang90) [inline etcdctlv3/command/error.go]: I think the cmd library we use will catch some of the bad values (like int flag but provided with hello string). I am not sure how to make this comment more accurate.

Comment (heyitsanthony) [inline etcdctlv3/command/error.go]: "unsupported" is more descriptive, I think; will change

Comment (xiang90) [inline etcdctlv3/command/error.go]: agree

**Pull request merged**